### PR TITLE
LibGUI: Propagate errors from the UndoStack

### PR DESCRIPTION
**Pull request by Baitinq**
* LibGUI+Userland: Propagate errors in GUI::*::{undo,redo}()
* PixelPaint: Refactor Image::restore_snapshot() for error handling

This removes a single FIXME :^)

### CODE REVIEW

**Comment by qeeg** [comment]
*Refactor

File: Userland/Applications/FontEditor/UndoSelection.h [hunk 1]
```diff
@@ -79,22 +80,28 @@ class SelectionUndoCommand : public GUI::Command {
...
+    virtual ErrorOr<void> undo() override
     {
         if (!m_redo_state) {
-            if (auto maybe_state = m_undo_state->save_state(); !maybe_state.is_error())
-                m_redo_state = move(maybe_state.value());
-            else
-                warnln("Saving redo state failed: {}", maybe_state.error());
+            auto save_state_or_error = m_undo_state->save_state();
+            if (save_state_or_error.is_error())
+                return Error::from_string_view(MUST(String::formatted("Saving redo state failed: {}", save_state_or_error.error())));
```
Comment (LucasChollet) [inline Userland/Applications/FontEditor/UndoSelection.h:88]: 1. You shouldn't use `MUST` in a fallible context (i.e. a function that returns an `ErrorOr`).
2. `StringView`s are non-owning. That means that creating a `StringView` from a temporary `String` will result in a new object that will never be correct, as it points to an already obsolete memory area. That's what you are currently doing with:
```cpp
Error::from_string_view(MUST(String::formatted("Saving redo state failed: {}", save_state_or_error.error())));
``` 
To resolve this, you can create a generic message and use `Error::from_string_literal()` or return the underlying error. The later can even be achieved with:
```cpp
auto save_state_or_error = TRY(m_undo_state->save_state());
```

File: Userland/Applications/FontEditor/UndoSelection.h [hunk 2]
```diff
@@ -79,22 +80,28 @@ class SelectionUndoCommand : public GUI::Command {
...
+        return {};
     }
-    virtual void redo() override
+    virtual ErrorOr<void> redo() override
     {
-        if (m_redo_state)
-            m_undo_selection.restore_state(*m_redo_state);
-        else
-            warnln("Restoring state failed");
+        if (!m_redo_state)
```
Comment (LucasChollet) [inline Userland/Applications/FontEditor/UndoSelection.h:99]: Using a `else` here is perfectly fine.

**Comment by LucasChollet** [comment]
Using a better PR name will probably bring more maintainer.
I find `LibGUI: Propagate errors from the UndoStack` more catchy :^)

**Comment by AtkinsSJ** [comment]
This has conflicts.

**Comment by Baitinq** [comment]
Bump

**Comment by AtkinsSJ** [comment]
Fixing the conflicts might be more useful. ;^)

**Pull request closed without merging**